crypto/pkcs7/pkcs7.c

```diff
@@ -608,3 +609,480 @@ void PKCS7_RECIP_INFO_get0_alg(PKCS7_RECIP_INFO *ri, X509_ALGOR **penc) {
+  GUARD_PTR(p7);
+  GUARD_PTR(md);
+  switch (OBJ_obj2nid(p7->type)) {
+    case NID_pkcs7_digest:
+      if (EVP_MD_nid(md) == NID_undef) {
+        OPENSSL_PUT_ERROR(PKCS7, PKCS7_R_UNKNOWN_DIGEST_TYPE);
+        return 0;
+      }
+      if (p7->d.digest->digest_alg) {
+        OPENSSL_free(p7->d.digest->digest_alg->parameter);
+      }
+      if ((p7->d.digest->digest_alg->parameter = ASN1_TYPE_new()) == NULL) {
+        OPENSSL_PUT_ERROR(PKCS7, ERR_R_ASN1_LIB);
+        return 0;
+      }
+      p7->d.digest->md = md;
+      p7->d.digest->digest_alg->parameter->type = V_ASN1_NULL;
+      p7->d.digest->digest_alg->algorithm = OBJ_nid2obj(EVP_MD_nid(md));
+      return 1;
+    default:
+      OPENSSL_PUT_ERROR(PKCS7, PKCS7_R_WRONG_CONTENT_TYPE);
+      return 0;
+  }
+}
+
+
+STACK_OF(PKCS7_RECIP_INFO) *PKCS7_get_recipient_info(PKCS7 *p7) {
+  GUARD_PTR(p7);
+  GUARD_PTR(p7->d.ptr);
+  switch (OBJ_obj2nid(p7->type)) {
+    case NID_pkcs7_enveloped:
+      return p7->d.enveloped->recipientinfo;
+    case NID_pkcs7_signedAndEnveloped:
+      return p7->d.signed_and_enveloped->recipientinfo;
+    default:
+      return NULL;
+  }
+}
+
+int PKCS7_dataFinal(PKCS7 *p7, BIO *bio) {
+  GUARD_PTR(p7);
+  GUARD_PTR(bio);
+  int ret = 0;
+  BIO *bio_tmp = NULL;
+  PKCS7_SIGNER_INFO *si;
+  EVP_MD_CTX *md_ctx = NULL, *md_ctx_tmp;
+  STACK_OF(PKCS7_SIGNER_INFO) *si_sk = NULL;
+  ASN1_OCTET_STRING *content = NULL;
+
+  if (p7->d.ptr == NULL) {
+    OPENSSL_PUT_ERROR(PKCS7, PKCS7_R_NO_CONTENT);
+    return 0;
+  }
+
+  md_ctx_tmp = EVP_MD_CTX_new();
+  if (md_ctx_tmp == NULL) {
+    OPENSSL_PUT_ERROR(PKCS7, ERR_R_EVP_LIB);
+    return 0;
+  }
+
+  switch (OBJ_obj2nid(p7->type)) {
+    case NID_pkcs7_data:
+      content = p7->d.data;
+      break;
+    case NID_pkcs7_signedAndEnveloped:
+      si_sk = p7->d.signed_and_enveloped->signer_info;
+      content = p7->d.signed_and_enveloped->enc_data->enc_data;
+      if (content == NULL) {
+        content = ASN1_OCTET_STRING_new();
+        if (content == NULL) {
+          OPENSSL_PUT_ERROR(PKCS7, ERR_R_ASN1_LIB);
+          goto err;
+        }
+        p7->d.signed_and_enveloped->enc_data->enc_data = content;
+      }
+      break;
+    case NID_pkcs7_enveloped:
+      content = p7->d.enveloped->enc_data->enc_data;
+      if (content == NULL) {
+        content = ASN1_OCTET_STRING_new();
+        if (content == NULL) {
+          OPENSSL_PUT_ERROR(PKCS7, ERR_R_ASN1_LIB);
+          goto err;
+        }
+        p7->d.enveloped->enc_data->enc_data = content;
+      }
+      break;
+    case NID_pkcs7_signed:
+      si_sk = p7->d.sign->signer_info;
+      // clang-format off
+OPENSSL_BEGIN_ALLOW_DEPRECATED
+      // clang-format on
+      content = PKCS7_get_octet_string(p7->d.sign->contents);
+      // clang-format off
+OPENSSL_END_ALLOW_DEPRECATED
+      // clang-format on
+      // If detached data then the content is excluded
+      // clang-format off
+OPENSSL_BEGIN_ALLOW_DEPRECATED
```

Comment:
NP: No need for the `OPENSSL_END_ALLOW_DEPRECATED` since we immediately need to allow it again.